Difficulty Bomb Delay to Summer 2022 (#3238)
ethereum/EIPs: EIPS/eip-3238.md

---
eip: 3238

> title: Vatnajoekull Difficulty Bomb Delay

Reviewer: Hard fork coordination happens separate from the EIP process in https://github.com/ethereum/eth1.0-specs/.  An EIP is just a technical document that describes a change, and has no concept of Ethereum mainnet fork names.

```suggestion
title: Difficulty Bomb Delay to Summer 2020
```

Author: OK but fixed the year ;-)

```suggestion
title: Difficulty Bomb Delay to Summer 2022
```

author: Afri Schoedon (@q9f)

> discussions-to: https://github.com/ethereum/EIPs/pull/3238

Reviewer: `discussions-to` cannot be the PR itself.  A thread over at https://ethereum-magicians.org/ is preferred, but if you don't have or want to create an account there you can create a in issue in this repository for "Discussion of ..." and link to it here.

Author: OK

```suggestion
discussions-to: https://github.com/ethereum/EIPs/issues/3239
```

type: Standards Track
category: Core
status: Draft
created: 2021-01-25
---

## Simple Summary

> The average block times are increasing due to the difficulty bomb (also known as the "_ice age_") and slowly accelerating. This EIP proposes to delay the difficulty bomb for approximately ten months with the Vatnajoekull fork, a yet to be scheduled adjustment.

Reviewer: This simply summary is a bit too long.  Think email subject line, GitHub pull request title, forum title, etc.

```suggestion
Delays the difficulty bomb so 30 second blocks won't happen until around summer 2022.
```

## Abstract

> Starting with `VTNJKLL_FORK_BLKNUM` the client will calculate the difficulty based on a fake block number suggesting to the client that the difficulty bomb is adjusting eleven million blocks later than previously specified with the Homestead fork; that is two million blocks later than specified with the latest delay as included in the Muir Glacier fork ([EIP-2387](./eip-2387.md)).

Reviewer: As mentioned above, EIPs are unrelated to hardfork coordination.  While you are free to choose any variable names you want, I recommend something simpler and more clear given the lack of context of hard fork naming in this repository like `FORK_BLOCK_NUMBER`.  (on a personal note, I also recommend spelling out words since you'll write this document once and it will be read thousands of times)

> Starting with `VTNJKLL_FORK_BLKNUM` the client will calculate the difficulty based on a fake block number suggesting to the client that the difficulty bomb is adjusting eleven million blocks later than previously specified with the Homestead fork; that is two million blocks later than specified with the latest delay as included in the Muir Glacier fork ([EIP-2387](./eip-2387.md)).

Reviewer: As mentioned above, try to keep the EIP purely technical.  Imagine someone is implementing this EIP for a chain that isn't Ethereum mainnet, which has no Homestead fork or Muir Glacier fork.  I also recommend against EIPs being *relative* changes, and instead be absolute changes.  So try to avoid statements like "+11M from previous".

```suggestion
Starting with `FORK_BLOCK_NUMBER` the client will calculate the difficulty based on a fake block number suggesting to the client that the difficulty bomb is adjusting eleven million blocks later than the actual block number.
```

## Motivation
Even after the Ethereum 2.0 mainnet launch, Ethash proof-of-work mining on the legacy chain should be feasible. It should allow miners sealing new blocks every 13~15 seconds on average for another ten months and allow both Ethereum 1.x and Ethereum 2.0 developers to conclude the merge.

## Specification
#### Relax Difficulty with Fake Block Number
For the purposes of `calc_difficulty`, simply replace the use of `block.number`, as used in the exponential ice age component, with the formula:

    fake_block_number = max(0, block.number - 11_000_000) if block.number >= VTNJKLL_FORK_BLKNUM else block.number

## Rationale
This will delay the ice age by another ~26 million seconds (approximately ~9.89 months), so the chain would be back at ~30 second block times in summer 2022. Hopefully, by then the Eth1-to-Eth2 merge will be concluded and the ice age fullfilled its task.

## Backwards Compatibility

> This EIP is not forward compatible and introduces backwards incompatibilities in the difficulty calculation. Therefore, it should be included in a scheduled hardfork at a certain block number. It's suggested to consider this EIP either with or shortly after the Berlin hard-fork but not later than July 2021.

Reviewer: Same comment about keeping hardfork coordination out of the EIP (don't mention Berlin).

Also, you can be a bit more terse about needing a hard fork if you want (not required).  It is generally assumed that anyone reading this will understand what a hard fork is so you can just say something like "this change will require a hard fork" or "this is a consensus change, so it will require a hard fork".

Author: If it's not a blocker, I would like to keep these considerations in the proposal. I understand they should work "out of context," but we have to admit this is a strong ethereum mainnet context even when we just say "Difficulty Bomb Delay to Summer 2022."

Reviewer: None of my comments on this section are a blocker for merging as a draft, but be warned that I probably won't merge a PR changing the status to Review as is.  Another editor might, but you'll have to shop around a bit.

We have been fighting a hard uphill battle getting hard fork coordination and governance stuff out of the EIP process and I am loath to give up this much ground in that battle, especially since it doesn't add any technical value to the EIP (it is purely governance discussion).

> Alternatively, in order to maintain stability of the system, a it can be considered to activate this EIP along with EIP-1559 fee market changes in a bundle. With the delay of the ice age, there is a desire to no further increase inflation and rather incentivize users to participate in proof-of-stake consensus instead. 

Reviewer: Recommend removing this whole paragraph as it is entirely hard fork coordination details (not appropriate for EIPs repository).
```suggestion
```

> ## Test Cases
> The test cases in their logic do not differ significantly from [EIP-649](./eip-649.md), [EIP-1234](./eip-1234.md), or [EIP-2384](./eip-2384.md) and can be duplicated.
> 

Reviewer: Consider either removing this section entirely, or specifying some tests in this document.  This is a weak suggestion, and if you think there is significant value to developers to keep this mention I think that is fine, but I'm of the weak opinion that it doesn't add much value and shorter EIPs tend to be preferable over longer ones.

```suggestion
```

> ## Implementation
> The implementation in its logic does not differ significantly from [EIP-649](./eip-649.md), [EIP-1234](./eip-1234.md), or [EIP-2384](./eip-2384.md) and should be implemented in all clients already.
> 

Reviewer: Consider removing this section for the same reason as above.

If you decide to retain this section, I recommend removing the "should be implemented in all clients already" as this document should be written for future readers, which includes clients that may not be written at the time they are reading this EIP.

```suggestion
```

> ## Copyright

Reviewer: Security Considerations section is required.

```suggestion
## Security Considerations
There are no known security issues with this proposal.

## Copyright
```
